### ISSUE TITLE
pylance incorrectly report AsyncGenerator of an abstract class.

### ISSUE THREAD

**Issue by Jacksunwei** [microsoft/pyright#9949]
## Environment data

-   Pylance version: 2025.2.1
-   OS and version: MacOS 15.3.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.11.11

## Code Snippet

```python
import abc
import asyncio
from typing import AsyncGenerator


class AbstractContentGenerator(abc.ABC):

  @abc.abstractmethod
  async def generate_content_async(
      self, input_str: str
  ) -> AsyncGenerator[str, None]:
    pass


class ContentGenerator(AbstractContentGenerator):

  async def generate_content_async(
      self, input_str: str
  ) -> AsyncGenerator[str, None]:
    for i in range(3):
      await asyncio.sleep(0.1)  # Simulate asynchronous operation
      yield f"Streaming content chunk {i+1} for: {input_str}"


async def test_1(generator: AbstractContentGenerator):
  async for response in generator.generate_content_async("input string"):
    print(f"Response (non-streaming): {response}")


async def main():
  generator: AbstractContentGenerator = ContentGenerator()
  await test_1(generator)


if __name__ == "__main__":
  asyncio.run(main())
```

## Repro Steps

1. XXX

## Expected behavior

pylance shouldn't report lint error.

## Actual behavior

![Image](https://github.com/user-attachments/assets/e97a21a9-8bff-4de2-9f55-4660d1229a0d)

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug.

The return type of an async function changes based on whether its body contains a `yield` statement. This mirrors the behavior at runtime. Without a `yield` statement in the body, the function becomes a coroutine rather than a generator, so static type checkers wrap the return type in `typing.Coroutine` or `types.CoroutineType`. Pyright and mypy are consistent in their behavior here.

To fix this in your code, you have a couple of options:
1. Add a yield statement to the abstract method body. You can place this after a `raise NotImplementedError` if you would like.
```python
class AbstractContentGenerator(abc.ABC):
    @abc.abstractmethod
    async def generate_content_async(self, input_str: str) -> AsyncGenerator[str, None]:
        raise NotImplementedError
        yield f""
```

2. Remove the `async` keyword from the abstract method. The `AsyncGenerator[str, None]` return type is sufficient to indicate that it's an async generator.
```python
class AbstractContentGenerator(abc.ABC):
    @abc.abstractmethod
    def generate_content_async(self, input_str: str) -> AsyncGenerator[str, None]:
        pass
```
